### PR TITLE
DR-899 Stairway Hooks

### PR DESCRIPTION
**Pull request by snf2ye**
Utilizing the model and infrastructure defined in stairway, these changes define and pass logging functions to Stairway. This will allow for logging to occur at the beginning and end of each flight and step. 
This will be merged in conjunction with [this PR.](https://github.com/DataBiosphere/stairway/pull/14)

### CODE REVIEW

File: src/main/java/bio/terra/service/job/JobService.java [hunk 1]
```diff
@@ -73,14 +77,50 @@ public JobService(IamService samService,
...
         stairway = Stairway.newBuilder()
             .maxParallelFlights(appConfig.getMaxStairwayThreads())
             .exceptionSerializer(serializer)
             .applicationContext(applicationContext)
             .stairwayName(appConfig.getPodName())
+            .stairwayHook(hooks)
         .build();
     }
 
+    private StairwayHook populateStairwayHooks(){
```
Comment (snf2ye) [inline src/main/java/bio/terra/service/job/JobService.java]: Is there a better place to put this chunk of code?

File: src/main/java/bio/terra/service/job/JobService.java [hunk 2]
```diff
@@ -73,14 +77,50 @@ public JobService(IamService samService,
...
+        return new StairwayHook() {
+            @Override
+            public HookAction startFlight(FlightContext context) {
+                logger.info("Operation: {}, flightClass: {}, flightId: {}, timestamp: {}",
+                    "startFlight", context.getFlightClassName(), context.getFlightId(), Instant.now());
```
Comment (snf2ye) [inline src/main/java/bio/terra/service/job/JobService.java]: Is this the correct way to grab the current timestamp for the log?

File: src/main/java/bio/terra/service/job/JobService.java (hunk 1 above)
Comment (ddietterich) [reply src/main/java/bio/terra/service/job/JobService.java]: I'd vote for making a separate derived class called `StairwayLoggingHooks` or some such, and then `new` it here. I am thinking that eventually, we may use some configuration setting to decide which of several hook implementations (or none) we want to create.

File: src/main/java/bio/terra/service/job/JobService.java (hunk 2 above)
Comment (ddietterich) [reply src/main/java/bio/terra/service/job/JobService.java]: Looks right to me.

File: src/main/java/bio/terra/service/job/JobService.java [hunk 3]
```diff
@@ -73,14 +77,50 @@ public JobService(IamService samService,
...
+                return HookAction.CONTINUE;
+            }
+
+            @Override
+            public HookAction startStep(FlightContext context) {
+                logger.info("Operation: {}, flightClass: {}, flightId: {}, stepIndex: {}, timestamp: {}",
```
Comment (ddietterich) [inline src/main/java/bio/terra/service/job/JobService.java]: minor: you could factor the strings into static final's.

**Review by ddietterich — APPROVED** [review]
LGTM - minor suggestions

File: src/main/java/bio/terra/service/job/JobService.java (hunk 2 above)
Comment (fboulnois) [reply src/main/java/bio/terra/service/job/JobService.java]: We might want to specify a timezone and formatting, something like this:

```suggestion
                    "startFlight", context.getFlightClassName(), context.getFlightId(), Instant.now().atZone(ZoneId.of("Z")).format(DateTimeFormatter.ISO_INSTANT));
```

**Review by fboulnois — COMMENTED** [review]
minor suggestion:

File: src/main/java/bio/terra/service/job/JobService.java (hunk 1 above)
Comment (snf2ye) [reply src/main/java/bio/terra/service/job/JobService.java]: This pattern makes a lot more sense, thank you!

**Review by ddietterich — APPROVED** [review]
LGTM

**Review by fboulnois — APPROVED** [review]
lgtm

**Pull request merged**